Turn the test type from a simple tuple into a record

mjambon opened the pull request:
I want to add support for capturing and checking the output of tests. We want the test framework to manage the output snapshots, show us the differences with expectations, and let us approve new output. This will require specifying options such as "capture stdout for this test and compare it against the expectation".

This PR creates a richer type for specifying tests and provides a function `create_test` with optional arguments. The legacy way of specifying a test using just a pair (name, function) is still supported and used in many places but it requires a call to `simple_test` or `simple_tests` to convert to the new type.

Future steps will involve implementing was I mentioned in https://github.com/mirage/alcotest/issues/400. If things work out, it may be integrated into Alcotest or become a standalone library that wraps around Alcotest. For now, we're keeping it in semgrep. The next steps are:
* figure out how to add subcommands to the test program (`status`, `approve`, `rerun`)
* implement stdout/stderr capturing
* implement the new subcommands

These initial changes will break semgrep-proprietary but it should only be a matter of inserting calls to `Testutil.simple_test` like I did on the semgrep code base. I'm not a fan of the name `Testutil`, I'm thinking of renaming it to something that alludes more to an experimental extension of Alcotest. Update: I renamed `Testutil` and isolated it in its own library named `Alcotest_ext`.

Commit:
Move away from representing tests as a tuple, allowing for various options.

Commit:
Fix more type errors

Commit:
Silence bad semgrep finding

Commit:
Move (most of) Testutil to Alcotest_ext

Commit:
Add missing files

Commit:
Fix type errors

aryx commented on dune-project:
```diff
@@ -28,10 +28,18 @@
 
 ; Opam packages
 
+(package
```
why a new package? Why not part of libs/commons/ like the other Arg_helpers.ml, Logger_helpers.ml, Cmdliner_helpers.ml (that we could rename to Xxx_ext.ml)

mjambon replied:
I'm determined to make this a standalone library that others can use as soon as possible, with the hope that some or all of it will be integrated into Alcotest eventually. For now, having it in its own folder without external dependencies other than Alcotest makes it easy for outsiders to review it.

aryx commented on libs/alcotest_ext/Alcotest_ext.ml:
```diff
@@ -0,0 +1,129 @@
+(* Helpers *)
+(****************************************************************************)
+(*
+   Dumb stuff for which
```
for which ???

mjambon replied:
🤔 it's probably a good idea that I didn't finish the sentence 😅. Something like the OCaml standard library should have had these functions 25 years ago.

aryx commented on libs/alcotest_ext/Alcotest_ext.ml:
```diff
@@ -0,0 +1,129 @@
+  (* Options *)
+  speed_level : Alcotest.speed_level;
+  check_output : output option;
+  (* Automatically determined *)
```
maybe explain here why we need an id for?

aryx commented on libs/alcotest_ext/Alcotest_ext.mli:
```diff
@@ -0,0 +1,141 @@
+   hesitate to log a lot during the execution of the test.
+*)
+type 'a t = private {
+  category : string list;
```
maybe add in comment an example of category and name, so people understand their use

aryx commented on libs/alcotest_ext/Alcotest_ext.mli:
```diff
@@ -0,0 +1,141 @@
+   Usage:
+
+     pack_tests_pro "Suite Name" [test1; test2]
+     pack_suites "Suite Name" [suite1; suite2; suite3]
```
maybe say that you can nest them, and that the name gets inserted in the category field of the test.

mjambon replied:
This is another thing we've doing on top of Alcotest (I thought I posted a feature request for this on https://github.com/mirage/alcotest/issues but I can't find it). The nesting is purely visual i.e. `["food"; "fruit"; "banana"]` becomes the test name `food > fruit > banana` and we can filter tests using the substring `od > fru`. It's weird but it doesn't hurt.

aryx commented on libs/commons/Testutil.mli:
```diff
@@ -1,142 +1,12 @@
-   hoisted all the way to the top level, so we can run them properly on the
-   JS runtime.
-   When running such tests in JS, we need our tests to also return promises.
+   Various utilities used for testing and are not considered an extension
```
Why not removing Testutil.ml instead of keeping some of it?

mjambon replied:
What's left in Testutil is what I don't think should go into the general-purpose library because it's Semgrep-dependent, it's not well thought out, or some other reason.

aryx reviewed (approved):
Looks good. Need to figure out why the JS tests are failing.

Commit:
Comments

Commit:
Add direct reference to alcotest_ext library

Commit:
Fix merge

mjambon commented on libs/aliengrep/Unit_Match.ml:
```diff
@@ -321,20 +321,21 @@ let test_caseless () =
+      ("right-anchored ellipses", test_right_anchored_ellipses);
+      ("pure ellipsis", test_pure_ellipsis);
+      ("caseless", test_caseless);
+    ]
```
Is this call at module initialization time a problem for semgrep js?

Commit:
Restore the original order of the libraries in a dune file in a
desperate attempt to avoid semgrep js failures.

mjambon commented:
https://github.com/semgrep/semgrep/pull/9343 moves the `Alcotest_ext` module back to the `commons` library, eliminating the `alcotest_ext` library. 
Let's see if it makes a difference.

Commit:
Update JS tests correctly without dropping options (using new update function)
This revises the interface of Alcotest_ext and introduces tags as a
means of filtering tests.

Commit:
Tag tests to be skipped by js tests

Commit:
Work around missing target files needed earlier than previously anticipated

Commit:
Comment

Commit:
Fix overly restrictive interface

Commit:
Skip more js tests

Commit:
Restrict tag syntax, print tags in Alcotest output

Commit:
Make the test tags abstract and prevent duplicate declarations

Commit:
Add a manual exclusion for a weird Ruby test, sort tags in output

Commit:
Fix bad rebase

Commit:
Fix OCaml formatting

Commit:
Fix bad merge

mjambon commented:
Ready to merge as soon as https://github.com/semgrep/semgrep-proprietary/pull/1129 is green.
